i2c/firmware/i2c.c

```diff
@@ -7,14 +8,40 @@
  */
 bool hal_i2c_master_transmit(HAL_I2C_HANDLE handle, uint16_t DevAddress, uint8_t *data, uint16_t size, uint32_t timeout)
 {
-    return HAL_I2C_Master_Transmit(handle,
-                            DevAddress, data, size, timeout) == HAL_OK;
+    I2C_HandleTypeDef* i2c_handle = (I2C_HandleTypeDef*)handle;
+    uint32_t tickstart = HAL_GetTick();
+    HAL_StatusTypeDef tx_result = HAL_OK;
+    do {
+        tx_result = HAL_I2C_Master_Transmit(i2c_handle,
+                            DevAddress, data, size, timeout);
+        if (__HAL_I2C_GET_FLAG(i2c_handle, I2C_FLAG_AF)) {
```

Comment:
Doesn't the `master_transmit` function set the busy flag already for you?

Reply:
this is the NACK (Ack Failure) flag not the busy flag